app/code/Magento/InventorySales/Plugin/CatalogInventory/StockManagement/ProcessRegisterProductsSalePlugin.php

```diff
@@ -81,9 +90,16 @@ public function aroundRegisterProductsSale(StockManagement $subject, callable $p
 
         $reservations = [];
         foreach ($productSkus as $productId => $sku) {
+            $qty = (float)$items[$productId];
+            $isSalable = $this->isProductSalableForRequestedQty->execute($sku, $stockId, $qty)->isSalable();
+            if (!$isSalable) {
+                throw new LocalizedException(
+                    __('Not all of your products are available in the requested quantity.')
+                );
+            }
```

Comment:
we need to make sure that  we can fulfill the order for all items and just after that place all the reservations.

So, you will have two loops (first - with isProductSalableForRequestedQty call  and all the preparation, second - create all the reservations )